Add SPIR-V 1.4 checks

MrSidims opened the pull request:
Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

MrSidims commented:
- [ ] Update tests

MrSidims commented on lib/SPIRV/libSPIRV/SPIRVValue.cpp:
```diff
@@ -102,11 +101,10 @@ void SPIRVValue::setNoUnsignedWrap(bool HasNoUnsignedWrap) {
   if (Module->isAllowedToUseExtension(
-          ExtensionID::SPV_KHR_no_integer_wrap_decoration)) {
+          ExtensionID::SPV_KHR_no_integer_wrap_decoration) &&
+      Module->isAllowedToUseVersion(VersionNumber::SPIRV_1_4)) {
```
Done

MrSidims commented on lib/SPIRV/libSPIRV/SPIRVValue.cpp:
```diff
@@ -80,11 +80,10 @@ void SPIRVValue::setNoSignedWrap(bool HasNoSignedWrap) {
   if (Module->isAllowedToUseExtension(
-          ExtensionID::SPV_KHR_no_integer_wrap_decoration)) {
+          ExtensionID::SPV_KHR_no_integer_wrap_decoration) &&
+      Module->isAllowedToUseVersion(VersionNumber::SPIRV_1_4)) {
```
done

MrSidims commented:
@Fznamznon @mlychkov @svenvh please take a look

MrSidims commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -3870,7 +3874,9 @@ bool LLVMToSPIRVBase::transExecutionMode() {
       case spv::ExecutionModeRoundingModeRTZ: {
-        if (!BM->isAllowedToUseExtension(ExtensionID::SPV_KHR_float_controls))
+        if (!(BM->isAllowedToUseExtension(
+                  ExtensionID::SPV_KHR_float_controls) &&
```
||

MrSidims commented on lib/SPIRV/libSPIRV/SPIRVValue.cpp:
```diff
@@ -78,13 +78,18 @@ bool SPIRVValue::hasNoSignedWrap() const {
 void SPIRVValue::setNoSignedWrap(bool HasNoSignedWrap) {
```
Done

mlychkov commented on lib/SPIRV/libSPIRV/SPIRVValue.h:
```diff
@@ -98,8 +98,31 @@ class SPIRVValue : public SPIRVEntry {
+      Module->setMinSPIRVVersion(
+          static_cast<SPIRVWord>(VersionNumber::SPIRV_1_4));
+      addDecorate(new SPIRVDecorate(NoIntegerWrapDecoration, this));
+      SPIRVDBG(spvdbgs() << "Set nsw for obj " << Id << "\n")
```
```suggestion
      SPIRVDBG(spvdbgs() << "Set "
                         << (NoIntegerWrapDecoration == DecorationNoSignedWrap) ? "nsw" : "nuw"
                         <<" for obj " << Id << "\n")
```

mlychkov commented on lib/SPIRV/libSPIRV/SPIRVValue.h:
```diff
@@ -98,8 +98,35 @@ class SPIRVValue : public SPIRVEntry {
+                   ExtensionID::SPV_KHR_no_integer_wrap_decoration)) {
+      Module->addExtension(ExtensionID::SPV_KHR_no_integer_wrap_decoration);
+      addDecorate(new SPIRVDecorate(NoIntegerWrapDecoration, this));
+      SPIRVDBG(spvdbgs() << "Set nsw for obj " << Id << "\n")
+    } else {
+      SPIRVDBG(spvdbgs() << "Skip setting nsw for obj " << Id << "\n")
```
Here as well. Maybe better to make a variable for that value?

MrSidims replied:
Applied

MrSidims commented:
There is an issue reported, which I can't reproduce locally. So I will debug through CI a bit, sorry for spam.

mlychkov commented on lib/SPIRV/libSPIRV/SPIRVValue.h:
```diff
@@ -98,8 +98,34 @@ class SPIRVValue : public SPIRVEntry {
+    // NoSignedWrap and NoUnsignedWrap decorations are available only if it is
+    // allowed to use SPIR-V 1.4 or if SPV_KHR_no_integer_wrap_decoration
+    // extension is enabled
+    const std::string InstStr =
```
The last moment here is that maybe it would be better to define InstStr only when SPIRVDBG is defined.

MrSidims replied:
Applied, thanks!

MrSidims commented:
I messed up with the git history a bit, changes from the last patch set: added macro and added spirv-max-version option=1.1 to two of the tests. Will create a translator issue in a moment.

MrSidims commented:
Created the following issue: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1205

mlychkov commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -3829,53 +3839,47 @@ bool LLVMToSPIRVBase::transExecutionMode() {
-              new SPIRVExecutionMode(BF, static_cast<ExecutionMode>(EMode))));
-          BM->addCapability(CapabilityKernelAttributesINTEL);
-        }
+        if (!BM->isAllowedToUseExtension(
```
Does this change functional or just a change of coding style?

MrSidims replied:
Just a coding style

mlychkov commented on lib/SPIRV/libSPIRV/SPIRVValue.h:
```diff
@@ -98,8 +98,36 @@ class SPIRVValue : public SPIRVEntry {
-  void setNoSignedWrap(bool HasNoSignedWrap);
-  void setNoUnsignedWrap(bool HasNoUnsignedWrap);
+
+  template <spv::Decoration NoIntegerWrapDecoration>
```
Just a nit: template function definition may be left in cpp file. One can just write explicit function specializations after the definition to eliminate linkage errors.

MrSidims replied:
Oh, I was trying to do that, but failed, but I just realized, that I tried it within a class, but should do specialization outside of the class definition.

Fznamznon commented on lib/SPIRV/libSPIRV/SPIRVValue.cpp:
```diff
@@ -124,6 +89,40 @@ void SPIRVValue::setFPFastMathMode(SPIRVWord M) {
 }
 
+template <spv::Decoration NoIntegerWrapDecoration>
+void SPIRVValue::setNoIntegerDecorationWrap(bool HasNoIntegerWrap) {
```
What is the purpose of HasNoIntegerWrap parameter? Calls to this functions are added only with true parameter and under if checking that the LLVM value has "NoIntegerWrap".

MrSidims replied:
I also find this strange, but prefer not to change it within this patch set (even if it already brings some code outside of the declared purpose).

Fznamznon replied:
:(

Fznamznon commented on test/transcoding/global_block.cl:
```diff
@@ -4,8 +4,18 @@
+//   %__block_literal_global = OpVariable %_ptr_CrossWorkgroup__struct_10
+//                         CrossWorkgroup %11
+// details can be found in:
+// – Public issue #35: OpEntryPoint must list all global variables in the
```
Is it a bug in the translator?

MrSidims replied:
Yes, see: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1205

mlychkov reviewed (approved):
LGTM, thank you.
